### PR TITLE
Solve error: ambiguous overload for ‘operator!=’ (operand types are ‘__half’ and ‘double’)

### PR DESCRIPTION
**Pull request by jiapei100**
**Remove the error: error: ambiguous overload for ‘operator!=’ (operand types are ‘__half’ and ‘double’) **

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@mshabunin could you take a look?

**Comment by mshabunin** [comment]
@jiapei100 , our CI CUDA builds have failed: https://github.com/opencv/opencv/actions/runs/5489317969/jobs/10008696594?pr=23948

**Comment by jiapei100** [comment]
@mshabunin 
Could be the reasons:

- `Ubuntu 20.04` vs. `Ubuntu 22.04`
- `Cuda 11.7` vs. `Cuda 12.2`


?????

BTW, are you building everything on your own **OpenCV CI** ?? Everything automatically?? 
That is cool...

File: modules/dnn/src/cuda4dnn/primitives/normalize_bbox.hpp [hunk 1]
```diff
@@ -111,7 +111,7 @@ namespace cv { namespace dnn { namespace cuda4dnn {
              * or there might be several weights
              * or we don't have to scale
              */
-            if (weight != (T)(1.0))
+            if (weight != static_cast(T)(1.0))
```
Comment (VadimLevin) [inline modules/dnn/src/cuda4dnn/primitives/normalize_bbox.hpp:114]: Wrong `static_cast` syntax, please refer to [cppreference](https://en.cppreference.com/w/cpp/language/static_cast) for more information about how it works.

```cpp
static_cast<T>(1.0)
```

**Comment by mshabunin** [comment]
Perhaps original issue is because of missing type case, but because of ambiguous constant type. Try to change to `1.f` and `0.f`.

**Comment by cudawarped** [comment]
> Perhaps original issue is no caused by missing type case, but by ambiguous constant type. Try to change to `1.f` and `0.f`.

The static cast fixes the issue with CUDA 12.2 but due to the ambiguous constant type you mentioned causes an issue with earlier versions of CUDA.

The combination of both, i.e.
```
if (weight != static_cast<T>(1.0f))
if(nms_iou_threshold > static_cast<T>(0.0f))
```
or double
```
if (weight != static_cast<T>(1.0))
if(nms_iou_threshold > static_cast<T>(0.0))
```
works on Windows with Visual Studio 2022 for CUDA 12.1 (old math api) and CUDA 12.2.

**Comment by Zacrain** [comment]
> > Perhaps original issue is no caused by missing type case, but by ambiguous constant type. Try to change to `1.f` and `0.f`.
> 
> The static cast fixes the issue with CUDA 12.2 but due to the ambiguous constant type you mentioned causes an issue with earlier versions of CUDA.
> 
> The combination of both, i.e.
> 
> ```
> if (weight != static_cast<T>(1.0f))
> if(nms_iou_threshold > static_cast<T>(0.0f))
> ```
> 
> or double
> 
> ```
> if (weight != static_cast<T>(1.0))
> if(nms_iou_threshold > static_cast<T>(0.0))
> ```
> 
> works on Windows with Visual Studio 2022 for CUDA 12.1 (old math api) and CUDA 12.2.

If I am not mistaken, ISO C++ is clear about default primitive types of constant literals. `1` defaults to `int` and can be implicitly type-casted when assigning to, e.g,, `double`, `1.0` defaults to `double`, `1.0f` to `float` and so on. So there should not be a "ambiguous constant type"
Therefore it shouldn't be necessary to specifically indicate the use of `double` or `float` when it is static_cast-ed to the target template type anyway. Maybe there are conversion functions missing for the "primitive" data types added by CUDA?
At this point I wonder whether this might be a problem of nvcc, although I couldn't find any info about that. Maybe that's intentional for some reason?

**Comment by cudawarped** [comment]
> If I am not mistaken, ISO C++ is clear about default primitive types of constant literals. 1 defaults to int and can be implicitly type-casted when assigning to, e.g,, double, 1.0 defaults to double, 1.0f to float and so on. So there should not be an "ambiguous constant type".
> Therefore it shouldn't be necessary to specifically indicate the use of double or float when it is static_cast-ed to the target template type anyway. Maybe there are conversion functions missing for the "primitive" data types added by CUDA?
> At this point I wonder whether this might be a problem of nvcc, although I couldn't find any info about that. Maybe that's intentional for some reason?

I think the use of the word ambiguous has itself become ambiguous!

I may be wrong but my interpretation of the build error

<details>
  <summary>Ubuntu2004-x64-CUDA / BuildAndTest output</summary>

> @jiapei100 , our CI CUDA builds have failed: https://github.com/opencv/opencv/actions/runs/5489317969/jobs/10008696594?pr=23948

> /home/ci/opencv/modules/dnn/src/layers/../cuda4dnn/primitives/region.hpp:124:37: error: call of overloaded '__half(int)' is ambiguous
>   124 |             if (nms_iou_threshold > (T)(0)) {
>       |                                     ^~~~~~
> In file included from /usr/local/cuda/include/cuda_fp16.h:3790,
>                  from /usr/local/cuda/include/cublas_api.h:75,
>                  from /usr/local/cuda/include/cublas_v2.h:65,
>                  from /home/ci/opencv/modules/dnn/src/layers/../cuda4dnn/csl/cublas.hpp:14,
>                  from /home/ci/opencv/modules/dnn/src/layers/../op_cuda.hpp:11,
>                  from /home/ci/opencv/modules/dnn/src/layers/region_layer.cpp:44:
> /usr/local/cuda/include/cuda_fp16.hpp:202:25: note: candidate: '__half::__half(double)'
>   202 |     __CUDA_HOSTDEVICE__ __half(const double f) { __x = __double2half(f).__x;  }
>       |                         ^~~~~~
> /usr/local/cuda/include/cuda_fp16.hpp:201:25: note: candidate: '__half::__half(float)'
>   201 |     __CUDA_HOSTDEVICE__ __half(const float f) { __x = __float2half(f).__x;  }
>       |                         ^~~~~~
> /usr/local/cuda/include/cuda_fp16.hpp:179:26: note: candidate: 'constexpr __half::__half(const __half&)'
>   179 | struct __CUDA_ALIGN__(2) __half {
>       |                          ^~~~~~

</details>

is that there is no conversion from `int` to `half` in CUDA 12.1 so it ambiguous as to whether it should cast the constant 0 to a `float` or a `double` as theses are the only two overloads available.

**Comment by cudawarped** [comment]
> @mshabunin
> Could be the reasons:
> 
>     Ubuntu 20.04 vs. Ubuntu 22.04
>     Cuda 11.7 vs. Cuda 12.2

@jiapei100 The CI uses an older version of CUDA, can you apply the changes suggested by @mshabunin  in addition to your changes, e.g.
```
if (weight != static_cast<T>(1.0f))
if (nms_iou_threshold > static_cast<T>(0.0f))
```
to this PR to see if it compiles successfully so we can update OpenCV to be compatilble with CUDA 12.2 please.

**Comment by ibrewster** [comment]
> 

For what it's worth, when I tried manually applying the suggested changes to my copy of the source on my Ubuntu 22.04 box with Cuda 12.2, it did resolve the errors being generated by those lines, however I ran into other issues later in compilation with different files (looks like the python bindings). Different issue, so not including the details unless requested...

**Comment by cudawarped** [comment]
@jiapei100 Kindly asking you to fix this PR so that OpenCV can support the latest version of CUDA.

**Comment by cudawarped** [comment]
@VadimLevin, @mshabunin is it possilble for the maintainers to apply the fixes discussed here (also shown [here](https://github.com/opencv/opencv/compare/4.x...cudawarped:opencv:cuda_fix_cuda_toolkit_12_2) on my forked branch) to see if this fixes the issue and passes all the relavent CI tests.  

CUDA 12.2 was released over a month ago with update 1 released yesterday and @jiapei100 seems to have gone awol.  It seems a shame to let this fix go stale when 🤞 it should be such a simple fix.

**Pull request closed without merging**